source/Halibut/Transport/Streams/AsyncStream.cs

```diff
@@ -0,0 +1,93 @@
+﻿#nullable enable
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Halibut.Util;
+
+namespace Halibut.Transport.Streams
+{
+    /// <summary>
+    ///     A Stream which helps with implementing everything that is required to make a stream async.
+    ///     This will force all required async methods to be implemented.
+    ///     This will also implement the old style async programing APM to go to the equivalent
+    ///     TAM (ie normal async) ReadAsync/WriteAsync method.
+    ///     This also ensures a async disposable method exists on the stream even in net48.
+    ///     net48 Stream does not implement IAsyncDisposable.
+    /// </summary>
+    public abstract class AsyncStream : AsyncDisposableStream
+    {
+        public sealed override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return await _ReadAsync(buffer, offset, count, cancellationToken);
+        }
+
+        protected abstract Task<int> _ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken);
```

Comment:
Nit: I love starting with a good underscore, but generally that's for member variables in .NET

Secondly, this goes against our style guide. As much as we would like to change it, I think it would be better to give these conforming names.

🤔 `ReadAsyncStream`? 🤷

Reply:
Yeah I tried doing something like that. In the end it resulted in clutter in the child classes. Clutter that no one cares about in the child classes. It isn't helpful to see `ReadAsyncStream`.

Reply:
@LukeButters Can we revisit the naming please as discussed Sync and @sburmanoctopus has also raised it. The _Method does not look as expected to the reader and in my opinion looks confusing and like a mistake in the name.

Reply:
PR to clean that up https://github.com/OctopusDeploy/Halibut/pull/455